vulkan: Add VK_NV_cooperative_matrix2 support for mul_mat and FlashAttention2

jeffbolznv opened the pull request:
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [ ] Medium
  - [x] High

This change adds support for VK_NV_cooperative_matrix2 (https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VK_NV_cooperative_matrix2, https://github.com/KhronosGroup/GLSL/blob/main/extensions/nv/GLSL_NV_cooperative_matrix2.txt) to the Vulkan backend. This is a recent Vulkan extension supported by NVIDIA drivers that enables matrix multiplies using the tensor cores, while being easier to use and supporting more operations than VK_KHR_cooperative_matrix.

While this PR is code complete and passes testing, it is a Draft for a while because the build system relies on the Vulkan SDK and the tooling for this extension will land in the next Vulkan SDK release (edit: available in SDK version 1.4.304.0). [out of date: In the meantime, if you're interested to try this out locally you can clone https://github.com/KhronosGroup/Vulkan-Headers and grab the latest CI build from https://github.com/google/shaderc and set the cmake variables Vulkan_INCLUDE_DIR to point to Vulkan-Headers/include and Vulkan_GLSLC_EXECUTABLE to point to glslc]. You'll need the most recent Vulkan beta driver from https://developer.nvidia.com/vulkan-driver.

The two main additions in this change are a coopmat2 mul_mat shader and a coopmat2 FlashAttention2 shader. The mul_mat shader supports normal matrix multiples and mixture of experts, and supports a variety of quantization formats using the "decode" callback functionality in coopMatLoadTensorNV. The decode callback functions are in dequant_funcs_cm2.comp and decode one element at a time. The FlashAttention2 shader also supports quantization formats and could theoretically use different formats for K and V, but the compilation cost for supporting all those combinations was too high and I don't know if this is ever used in practice. Note that the mul_mat approach to quantization formats is analogous to the existing Vulkan mul_mat shader in that it converts to fp16 before multiplying, whereas I believe the CUDA path converts to int8 and applies the scale/bias per-tile.

I've also done optimizations of some other shaders, including mul_mat_vec, split_k_reduce, and adding a "linear vec4 copy" shader. With the much higher cooperative matrix2 perf for mul_mat, these other state buckets become relatively more expensive. I'll split these out into smaller changes to review separately. But I wanted to include them all here for context and for perf measurements.

[Edit: Previous performance comparisons against existing Vulkan path were broken. The comparison between Vulkan w/coopmat2 and CUDA I believe was accurate.]

The coopmat2 path helps significantly with prompt processing. FA helps a little bit in token gen, but only a few percent here and there.


![rtx4070pp512-fixed](https://github.com/user-attachments/assets/c473e02d-a6d2-4aff-8850-7d3b8690ef90)
![rtx4070tg128-fixed](https://github.com/user-attachments/assets/bb378acd-38a5-4e1f-9aae-20f58cd7fe97)
![rtx6000pp512-fixed](https://github.com/user-attachments/assets/b65a7a54-5774-4be7-b3d1-3e44ecedb270)
![rtx6000tg128-fixed](https://github.com/user-attachments/assets/64f15dc5-3648-40ae-b5cb-02b3d1a84e3f)

jeffbolznv committed:
vulkan: Add VK_NV_cooperative_matrix2 support for mul_mat and flash attention

0cc4m commented:
Wow, this looks really impressive. I'll take a closer look later.

In the interest of compatibility, how much work do you think it is to also add a VK_KHR_cooperative_matrices codepath, for AMD RDNA3 and maybe Intel ARC? Obviously not something you have to do, but I'd have to try it sooner or later.

jeffbolznv commented:
> how much work do you think it is to also add a VK_KHR_cooperative_matrix codepath

It may not be too bad in the mul_mm shader since you already have the code to dequantize and copy to shared memory. Porting the flash attention shader would be more involved, both because there's no existing scalar shader to start from, and also because it does reductions and needs to clear padding elements, and there's no way to do that other than to dump the matrix out to shared memory.

lin72h commented:
Very impressive. Just to let you know, Mesa's amdgpu driver is[ implementing a polyfill](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/31112) for pre-RDNA3 generation, so VK_KHR_cooperative_matrix optimization could benefit all AMD GPUs running Linux.

0cc4m commented:
> Very impressive. Just to let you know, Mesa's amdgpu driver is[ implementing a polyfill](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/31112) for pre-RDNA3 generation, so VK_KHR_cooperative_matrix optimization could benefit all AMD GPUs running Linux.

Probably not, emulating the matrix operations will likely just be slower than a non-coopmat codepath. You can see an example of that at the bottom of the merge request.

0cc4m commented:
There's more parts to this that could be upstreamed separately, before the driver and Vulkan support for the coopmat2 extension lands. I'm thinking of the vector copy shader and the matrix vector multiplication shader improvements.

Additionally, if we upstream the f16acc and f32acc switch earlier we could adapt my matrix multiplication shader to it. I had previously hardcoded the accumulator to float due to precision issues.

In the meantime I'll look into a `VK_KHR_cooperative_matrix` implementation in the regular matmul shader.

jeffbolznv commented:
I'm currently working on the optimized copy shader and will make a PR soon. I had trouble reproducing the gains from the mat-vec mul shader in isolation, but I think it may have only really benefited Q8_0. I'll try that again soon.

If you're going to look into the KHR path, I did a very basic prototype of it recently that you could use as a starting point: https://github.com/jeffbolznv/llama.cpp/commit/3416010. The first real bit of nastiness I ran into is needing to bound-check the store (see the comment).

netrunnereve commented:
@jeffbolznv do you happen to know approximately how much of the tg improvement is due to tensor cores and how much is due to flash attention? Dedicated matrix matrix multipliers probably are going to help a lot with prompt processing but the matrix vector multiplications for interence are limited by memory bandwidth in most cases. From the graphs it looks like the original Vulkan implementation is severely compute bound.

jeffbolznv commented:
Oh no, I'm afraid there was a mistake in our testing methodology where we had `-fa 1` set for the existing Vulkan path, and it causes the FA nodes to quietly fallback to the CPU. There's little gain from the cooperative matrix 2 path for token generation (likely a few percent from FA). Still a good gain for prompt processing, but those results are also heavily skewed. I'm going to edit the first comment to retract them and I'll try to update with more accurate results in the next day or so. My apologies, everybody.

jeffbolznv commented:
I've replaced the RTX 4070 results with something I think is more correct. Very little gain for tg, still large gain for pp. Will try to get updated RTX 6000 results tomorrow.

ggerganov commented:
> Very little gain for tg, still large gain for pp. Will try to get updated RTX 6000 results tomorrow.

FA for TG starts to make significant difference only at large contexts so this is expected for the `tg128` test. You can use the `llama-batched-bench` tool to estimate the TG speedup at larger contexts (see the OP in https://github.com/ggerganov/llama.cpp/pull/10171 for the commands).

0cc4m commented:
Does someone have a theory for the DeepSeek Coder V2 Lite outlier in the benchmarks? Why is Vulkan significantly outperforming CUDA there, in tg even without the tensor cores. Some issue with the MUL_MAT_ID implementation?

slaren commented:
Possibly because the expert selection is done on the CPU.

netrunnereve commented:
So I tried out the `mul_mat_vec.comp` changes only and got 6% faster inference speed with Q4_0. Pretty much all the improvements came from the barrier free subgroup add, and the additional unrolling and optimizations didn't seem to help much. Interestingly enough I got a negligible difference in performance when I tried to implement the subgroup add in the separate K-quant mat mul shaders.

BTW our mat vec shaders have a subgroup size of 32 which is fine for Nvidia and new AMD cards but bad for my old W8100 :smirk:.

jeffbolznv commented:
I'm working on splitting out the mul_mat_vec changes and making some additional optimizations, I hope to make a PR for that tomorrow. I haven't actually seen much gain from the subgroupAdd, and using that if the subgroup size is not equal to 32 is tricky, so I wasn't planning to leave that in. Was it on W8100 that you saw it helped?

0cc4m commented:
I haven't used the subgroup operations so far cause they caused driver crashes on Intel. But I'll test that again once you make the PR. If that problem is still around we might have to make them optional.

GCN cards are definitely relevant, as for many of them Vulkan is the last API they have left. But also modern RDNA defaults to subgroups of size 64, though you can manually reduce that to 32 using `VK_EXT_subgroup_size_control`. Currently that is not done. I would like to support subgroups of 32 and 64 where possible.

sorasoras commented:
> I haven't used the subgroup operations so far cause they caused driver crashes on Intel. But I'll test that again once you make the PR. If that problem is still around we might have to make them optional.
> 
> GCN cards are definitely relevant, as for many of them Vulkan is the last API they have left. But also modern RDNA defaults to subgroups of size 64, though you can manually reduce that to 32 using `VK_EXT_subgroup_size_control`. Currently that is not done. I would like to support subgroups of 32 and 64 where possible.

I wonder what kind of perf would be if rdna3 run at dual issue 32wave vopd.

https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Ff1c6cf47-ab88-47b7-a377-ef733a1b6c4b_670x429.png

netrunnereve commented:
> I haven't actually seen much gain from the subgroupAdd, and using that if the subgroup size is not equal to 32 is tricky, so I wasn't planning to leave that in. Was it on W8100 that you saw it helped?

Yep I got a 6% improvement on my W8100 by replacing the final barrier sum with something like this. I guess it might be due to the fact that my card is slower at thread synchronization and gets hung up by the two barriers, but I don't know.
```
    // sum up partial sums and write back result
    tmp = subgroupAdd(tmp);
    if (tid == 0) {
        data_d[d_offset + row] = D_TYPE(tmp);
    }
```

Honestly I don't really care about a 6% change for quants that I don't use anyways, so don't fret about this. What I'll probably look into instead is the 64 subgroup size as my card is potentially only using half its capabilities with 32 and that should hopefully have a big improvement. The shaders look pretty simple to hack on as long as I don't need to touch `ggml-vulkan.cpp`.

JohannesGaessler commented:
The Vulkan performance vs. CUDA with this PR is definitely much better than I would have expected would be possible using a more generic API. I'll benchmark the performance myself once the feature becomes available via package managers.

>The decode callback functions are in dequant_funcs_cm2.comp and decode one element at a time.

This is an on-the-fly dequantization to FP16 into SRAM, right? Context: llama.cpp training support is one of my next goals and I think memory use will be the biggest bottleneck. So good performance for small batch sizes would be desirable and dequantization into VRAM would be too slow I think.

>The FlashAttention2 shader also supports quantization formats and could theoretically use different formats for K and V, but the compilation cost for supporting all those combinations was too high and I don't know if this is ever used in practice.

The K cache seems to need more precision than the V cache so an asymmetric setup could make sense for users that like fiddling.

>Note that the mul_mat approach to quantization formats is analogous to the existing Vulkan mul_mat shader in that it converts to fp16 before multiplying, whereas I believe the CUDA path converts to int8 and applies the scale/bias per-tile.

That is correct, the CUDA code uses either `__dp4a` or int8 tensor cores. I did this under the expectation that I would be able to get better throughput than FP16 cuBLAS GEMM but for the kernel itself (and ignoring any auxiliary kernels) I was unfortunately not able to get better performance. I think the reason is the comparatively small block size of 16/32 of the current GGML quantization formats. I want to at some point try making a format specifically for directly training as quantized where I think it would make sense to use larger blocks for better performance.

>FA helps a little bit in token gen, but only a few percent here and there.

Try setting `-n 8192`, you need a large enough context for the results to be sensitive to the performance of the attention.

jeffbolznv commented:
> > The decode callback functions are in dequant_funcs_cm2.comp and decode one element at a time.

> This is an on-the-fly dequantization to FP16 into SRAM, right? 

Yes, the intent is that the compiler should be staging the matrices through shared memory. That may not happen in all cases, depends on the implementation, but that's the goal.

0cc4m commented:
> If you're going to look into the KHR path, I did a very basic prototype of it recently that you could use as a starting point: [jeffbolznv@3416010](https://github.com/jeffbolznv/llama.cpp/commit/3416010). The first real bit of nastiness I ran into is needing to bound-check the store (see the comment).

Thank you @jeffbolznv. I also had a basic non-functional version I could start out with. I compared them and built something that looks mostly correct to me, but I can't seem to get correct results and I'm running out of ideas what the issue is.

I've uploaded the current state to https://github.com/ggerganov/llama.cpp/tree/0cc4m/vulkan-coopmat / f5cae8d193fddc6bec43539a44cd7149ecba042c. If built with `GGML_VK_RUN_TESTS` it runs some small basic matrix multiplications and checks the results.

It seems to be some issue with loading the coopmats, but it's very hard to debug. I think at least the partial store is correct. The load does not seem to work as I expect, since I get the results I want only when I stage the load from `buf_a` through `coopmat_stage` shared memory and then into the coopmat, not if I load directly from `buf_a`. Some of what I tried is commented out in the file.

I'm using an RTX 3090 to test it. Let me know if you see something. This is kinda off-topic for this PR though, I don't mean to hijack it for something else.

jeffbolznv commented:
> I'm using an RTX 3090 to test it. Let me know if you see something. This is kinda off-topic for this PR though, I don't mean to hijack it for something else.

Looks like each warp is only computing one tile of the result (WMxWN = 16x16), which doesn't line up with the wgdenoms/workgroupsize it's using. I think that at least explains the first failing test (32x32x16).

0cc4m commented:
> Looks like each warp is only computing one tile of the result (WMxWN = 16x16), which doesn't line up with the wgdenoms/workgroupsize it's using. I think that at least explains the first failing test (32x32x16).

I haven't looked into that yet because my assumption was that it should work for matrices small enough, like the 4x4x4 test I was using. Only one coopmat multiplication is required for that.

jeffbolznv commented:
Strange, 4x4x4 is passing for me, the first failure I saw was in 32x32x16.

0cc4m commented:
> Strange, 4x4x4 is passing for me, the first failure I saw was in 32x32x16.

Okay, that's odd. Implementation difference maybe? You're using Ada, right, not Ampere?

jeffbolznv commented:
> Okay, that's odd. Implementation difference maybe?

I wonder if this test doesn't have the stride a multiple of 16B. That's required by the spec, but I think recent drivers with the coopmat2 implementation made the driver more forgiving about this.

0cc4m commented:
> > Okay, that's odd. Implementation difference maybe?
> 
> I wonder if this test doesn't have the stride a multiple of 16B. That's required by the spec, but I think recent drivers with the coopmat2 implementation made the driver more forgiving about this.

Oh yeah, that must be it. I'm running into undefined behaviour cause the stride isn't aligned. `buf_a` and `buf_b` use a stride of `BK + 1` to avoid bank conflicts, and that would be 17 in my first tests. I missed that detail in the spec, it would have probably taken me a long time to find that. Thank you!

0cc4m commented on ggml/src/ggml-vulkan/vulkan-shaders/dequant_funcs_cm2.comp:
```diff
@@ -0,0 +1,305 @@
+
+#include "types.comp"
+
+layout(buffer_reference, std430, buffer_reference_align = 2) buffer decodeBufQ4_0 {
```
Is buffer_reference supported without any device extension or minimum Vulkan API version requirement?

jeffbolznv replied:
There's a feature enable for it in VkPhysicalDeviceVulkan12Features, so it's getting filled out when it's queried from the driver.

0cc4m replied:
I see, might be good to check for it. It seems most drivers support it at this point, except some phones and SoCs.

jeffbolznv replied:
will do

0cc4m commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -2774,6 +3107,16 @@ static vk_pipeline ggml_vk_guess_matmul_pipeline(ggml_backend_vk_context * ctx,
+        if ((m % mmp->m->wg_denoms[0]) == 0 && (n % mmp->m->wg_denoms[1]) == 0) {
+            return aligned ? mmp->a_m : mmp->m;
+        }
+        return aligned ? mmp->a_s : mmp->s;
```
Interesting. I guess the logic here is to only use larger tiles if this doesn't require padding? Do you think this is applicable to the generic shader too, or specific to your cm2 version?

jeffbolznv replied:
Right, it's trying to use the largest size that doesn't require bounds-checking, because the bounds-checking is moderately expensive. In the generic shader I think the bounds checking is a much smaller impact and is currently always enabled, so similar logic probably wouldn't be helpful without also adding non-bounds-checking variants.

0cc4m replied:
No, bounds checking only happens on the non-aligned f16 and f32 loads. For the vectorized and quant loads no checks are done.

jeffbolznv replied:
No, you're right, I was mistaken and though there were bounds checks on the loads. Since you're already getting the benefit of not bounds checking, then this logic probably wouldn't help for the existing shaders.

0cc4m commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -2937,18 +3280,20 @@ static void ggml_vk_mul_mat_q_f16(ggml_backend_vk_context * ctx, vk_context& sub
-    const bool y_non_contig = !ggml_vk_dim01_contiguous(src1);
+    // Reformat and convert to fp16 if src1 is non-contiguous, or for coopmat2 for better perf
+    const bool y_non_contig = (ctx->device->coopmat2 && src1->type == GGML_TYPE_F32) ||
+                              !ggml_vk_dim01_contiguous(src1);
```
Do you know why this leads to better performance? I assumed it's better to avoid another buffer and copy shader call if possible. Is this maybe also applicable to the non-cm2 shader?

jeffbolznv replied:
With coopmat2 the shaders become very limited by the memory system and it's worth it to convert f32->f16 to reduce the amount of data that needs to be loaded. It may also help the coopmat1 path. I doubt it would have a big impact on the non-coopmat path.

0cc4m commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -2729,15 +3048,29 @@ static void ggml_vk_buffer_memset(vk_buffer& dst, size_t offset, uint32_t c, siz
+                split_k = 2;
+            }
+        }
+    }
```
Assuming split_k works reliably here, I guess the logic should also work for the non-cm2 codepath. I don't quite remember what issues specifically I had with it to make me disable it. I guess I should look into it again after this change is merged.

jeffbolznv replied:
There was a bug where split_k > 4 didn't work, the fix is at line 3338. Otherwise I didn't run into any functional issues. Whether it helps performance is still very system-dependent and we'd need to do testing to enable it more broadly.

0cc4m commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -6791,7 +7346,10 @@ static const struct ggml_backend_device_i ggml_backend_vk_device_i = {
+
+    const char *name = getenv("GGML_VULKAN_DISABLE_CM2") ? (GGML_VK_NAME " (disable CM2)") : GGML_VK_NAME;
+
+    return name;
```
Debugging

0cc4m commented on ggml/src/ggml-vulkan/vulkan-shaders/mul_mat_split_k_reduce.comp:
```diff
@@ -5,25 +5,44 @@
 layout (binding = 0) readonly buffer A {float data_a[];};
+layout (binding = 0) readonly buffer A4 {vec4 data_a4[];};
 layout (binding = 1) writeonly buffer D {float data_d[];};
+layout (binding = 1) writeonly buffer D4 {vec4 data_d4[];};
```
This double binding is very cool, I wasn't aware this was possible.

jeffbolznv replied:
Yes, it's legal, the spec even calls it out in a note:

```
If multiple shader variables are declared with the same set and binding
values, and with the same underlying descriptor type, they can all be
statically used within the same shader.
```

0cc4m commented on ggml/src/ggml-vulkan/vulkan-shaders/mul_mm_cm2.comp:
```diff
@@ -0,0 +1,328 @@
+        stride_b &= ~7;
+#if QUANT_K == 1
+        stride_a &= ~7;
+#endif
```
Do these kinds of hints make a difference? Quite interesting.

jeffbolznv replied:
The coopmat2 part of the NV compiler needs to detect alignment in order to use larger loads (and potentially other optimizations), so it does help. This sort of thing probably doesn't help non-coopmat2 code, though maybe it could if it helped vectorization?

0cc4m commented on ggml/src/ggml-vulkan/vulkan-shaders/mul_mm_cm2.comp:
```diff
@@ -0,0 +1,328 @@
+
+    // Convert from ACC_TYPE to D_TYPE
+    coopmat<D_TYPE, gl_ScopeWorkgroup, BM, BN, gl_MatrixUseAccumulator> mat_d;
+    mat_d = coopmat<D_TYPE, gl_ScopeWorkgroup, BM, BN, gl_MatrixUseAccumulator>(sum);
```
Do you know if this kind of conversion is possible with KHR coopmat extension? I've run into issues with trying to load or store when the types weren't equal.

jeffbolznv replied:
Yeah, with coopmat1 it's possible to convert to a matrix that only differs by component type. With coopmat2 it's also possible to convert to a matrix with a different MatrixUse.

0cc4m reviewed (commented):
Apologies for the number of comments.

jeffbolznv commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -1662,8 +1873,8 @@ static vk_device ggml_vk_get_device(size_t idx) {
-        // Try to find a non-graphics compute queue and transfer-focused queues
-        const uint32_t compute_queue_family_index = ggml_vk_find_queue_family_index(queue_family_props, vk::QueueFlagBits::eCompute, vk::QueueFlagBits::eGraphics, -1, 1);
+        // Try to find a queue that supports compute and transfer-focused queues
+        const uint32_t compute_queue_family_index = ggml_vk_find_queue_family_index(queue_family_props, vk::QueueFlagBits::eCompute, vk::QueueFlags{}, -1, 1);
```
It wasn't wrong, I think performance will be the same for either type of queue. I can change it back if you want.

0cc4m replied:
Let's just keep it as is then, it probably doesn't make a difference, but I'm not sure on AMD.

jeffbolznv replied:
done.

jeffbolznv commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -1728,6 +2027,9 @@ static vk_device ggml_vk_get_device(size_t idx) {
         device->name = GGML_VK_NAME + std::to_string(idx);
+        if (getenv("GGML_VULKAN_DISABLE_CM2")) {
+            device->name += " (disable CM2)";
+        }
```
Yeah, the name stuff is just to make it easier for me to generate graphs with both sets of results. I'll remove it when we get closer to merging.

jeffbolznv commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -1707,12 +1918,100 @@ static vk_device ggml_vk_get_device(size_t idx) {
         device->pipeline_robustness = pl_robustness_features.pipelineRobustness;
 
+        // XXX TODO remove env var?
+        if (!getenv("GGML_VULKAN_DISABLE_CM2")) {
```
will do

0cc4m commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -1320,7 +1421,101 @@ static void ggml_vk_load_shaders(vk_device& device) {
+        CREATE_MM(pipeline_dequant_mul_mat_mat_f16[GGML_TYPE_Q4_K].f16acc, matmul_q4_k_f16, _f16acc, mmq_wg_denoms_k, warptile_mmq_k, vk_mat_mat_push_constants, 3)
+        CREATE_MM(pipeline_dequant_mul_mat_mat_f16[GGML_TYPE_Q5_K].f16acc, matmul_q5_k_f16, _f16acc, mmq_wg_denoms_k, warptile_mmq_k, vk_mat_mat_push_constants, 3)
+        CREATE_MM(pipeline_dequant_mul_mat_mat_f16[GGML_TYPE_Q6_K].f16acc, matmul_q6_k_f16, _f16acc, mmq_wg_denoms_k, warptile_mmq_k, vk_mat_mat_push_constants, 3)
+        CREATE_MM(pipeline_dequant_mul_mat_mat_f16[GGML_TYPE_IQ4_NL].f16acc, matmul_iq4_nl_f16, _f16acc, mmq_wg_denoms, warptile_mmq, vk_mat_mat_push_constants, 3)
```
Yeah, it's plausible. Why didn't you do the same to the mul_mat_id shaders?

jeffbolznv replied:
No good reason. I didn't focus as much on mul_mat_id because I didn't have as many test cases using it.

jeffbolznv commented:
I was originally thinking we'd not merge this until the next Vulkan SDK is released, but I recently realized that Android has a separate SDK on a different release schedule, so the code needs to build against older headers for a while regardless. I've updated things so it can build against older or newer headers. For the shader compiles, it checks the vulkan header for the presence of the extension and assumes glslc will support coopmat2 if the extension is present in the headers. I don't know of a better way to check for support.

I've removed the debug code and will remove the "draft" label, I think this is ready for review/merge. I'm fine with waiting for #10597 to be merged first.

slaren commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -6751,6 +7242,46 @@ static bool ggml_backend_vk_device_supports_op(ggml_backend_dev_t dev, const ggm
+        case GGML_OP_FLASH_ATTN_EXT:
+            {
+                ggml_backend_vk_device_context * ctx = (ggml_backend_vk_device_context *)dev->context;
+                if (!ggml_vk_get_device(ctx->device)->coopmat2) {
+                    return false;
+                }
+                if (op->src[0]->type != GGML_TYPE_F32) {
+                    return false;
+                }
+                if (op->type != GGML_TYPE_F32) {
+                    return false;
+                }
+                if (op->src[3] && op->src[3]->type != GGML_TYPE_F16) {
+                    return false;
+                }
+                // It's straightforward to support different K/V dequant, but would
+                // significantly increase the number of pipelines
+                if (op->src[1]->type != op->src[2]->type) {
+                    return false;
+                }
+                switch (op->src[1]->type) {
+                case GGML_TYPE_F16:
+                case GGML_TYPE_Q4_0:
+                case GGML_TYPE_Q4_1:
+                case GGML_TYPE_Q5_0:
+                case GGML_TYPE_Q5_1:
+                case GGML_TYPE_Q8_0:
+                // K dequants currently disabled because D dimension is rounded up to 256 and runs inefficiently
+                //case GGML_TYPE_Q2_K:
+                //case GGML_TYPE_Q3_K:
+                //case GGML_TYPE_Q4_K:
+                //case GGML_TYPE_Q5_K:
+                //case GGML_TYPE_Q6_K:
+                case GGML_TYPE_IQ4_NL:
+                    break;
+                default:
+                    return false;
+                }
+                return true;
+            }
```
Since only a few values are supported, add a check here to determine if the head dimension is supported.

jeffbolznv replied:
done, thanks.

0cc4m commented:
> I've removed the debug code and will remove the "draft" label, I think this is ready for review/merge. I'm fine with waiting for #10597 to be merged first.

I'll do a review in a few hours. If no bigger issues come up, I think we can merge this one first, afterall my PR was inspired by this one.

0cc4m commented:
I managed to update my libvulkan, the Vulkan headers and glslc, so that everything compiles. But I don't have the coopmat2 extension on my RTX 3090 despite using driver `565.57.01`. According to Nvidia's [Vulkan Driver Support](https://developer.nvidia.com/vulkan-driver) page it should be included since `Linux 550.40.79`. Am I misunderstanding how Nvidia driver versions work or what is going on here? I'd rather avoid the `.run` files if necessary, but I'm already on the latest version available from Nvidia's repo.

jeffbolznv commented:
The Vulkan Developer Beta Drivers are based on a slightly older branch than the current General Release drivers, and so they have a lower version number, but it's the branch we use to release new features and bugfixes for developers without the longer release latency of the general release drivers. So for now, the coopmat2 support is only in the developer beta drivers.

0cc4m commented:
I managed to install it. It works well, a decent amount faster than my coopmat implementation. I tested it on Linux RTX 3090:

![coopmat_rtx3090_pp](https://github.com/user-attachments/assets/382d7c32-9a91-458e-a018-99c9fbf3ca81)
![coopmat_rtx3090_tg](https://github.com/user-attachments/assets/6fe79aa6-8b09-4bc6-ae15-9f81386e985f)

0cc4m reviewed (approved):
Looks good. It compiled and ran fine on an older driver without the feature and on an older Vulkan version (but obviously didn't give any advantage there).

jeffbolznv commented:
The latest Vulkan SDK (1.4.304.0) includes support for cooperative matrix 2. A developer driver from https://developer.nvidia.com/vulkan-driver is still necessary.

etasnadi commented:
Hi, I have recently been informed that this project might needs FlashAttention-2 using `VK_KHR_cooperative_matrix`. Fortunately, I have an implementation for it and I'm willing to merge it if it is still the case. 

The only problem that it peaks at 30% of the theoretical flops limit (at least on my RTX 2060 super) with fp16. It might be further optimized or not -- in theory, KHR coopmat can be significantly slower than NV_coopmat as the former does not allow us to perform complex manipulation of the matrices while they reside in registers, so my alg needs to write intermediate results to the shared memory and read back again several times.

See the code here: https://github.com/etasnadi/VulkanCooperativeMatrixAttention .

I also have a blogpost that summarizes the results: https://etasnadi.com/2025/01/vulkanglsl-implementation-of-the-scaled-dot-product-attention-sdpa-with-vk_khr_cooperativematrix/

0cc4m commented:
Hi @etasnadi, I wasn't aware that someone already implemented this, that is very cool! Of course we could use that, I'm happy to work with you to modify your work to run in llama.cpp. One step would be to implement support for quantized weight matrices.

Let me know how you want to proceed.

etasnadi commented:
Hi, that's a good idea. 

I spent some time with this project in the weekend and I managed to enable coopmat2 on my system so far.

Recently I investigated the scheduling logic, because I observed that the flash attention op actually scheduled to run on CPU (instead with coopmat2) because the models I downloaded are not compatible with Jeff's op or flash attention in general.

If you can provide some information about the internals that would significantly reduce the development time because now I have to figure out how things work under the hood. 

Where does this kind of communication happens in the project. Does the project have a slack/discord channel or something?

JohannesGaessler commented:
My communication with other devs is 99% Github and 1% email.

0cc4m commented:
> Where does this kind of communication happen in the project. Does the project have a slack/discord channel or something?

Most of the communication happens on Github, but in this case I can help you more directly on Discord, if you want. You can add me there (`_occam`)

etasnadi commented:
Thanks. I am more comfortable with Github and e-mail too, so if the communication happens here, I am not opening a slack/discord account.

0cc4m commented:
Fair enough. Then I suggest you open an issue or draft PR for this topic and put your questions there and I'll do my best to help.

jeffbolznv commented:
The r575 game ready drivers supporting NV_cooperative_matrix2 are available now, e.g. version 576.02 for Windows. You may also see a small boost in token generation perf.
